Rewriting parts of the README and getting started guides for 4.0 (#6767)
gradio-app/gradio: guides/01_getting-started/01_quickstart.md

## Building Your First Demo

You can run Gradio in your favorite code editor, Jupyter notebook, [Google Colab](https://colab.research.google.com/drive/18ODkJvyxHutTN0P5APWyGFO_xwNcgHDZ?usp=sharing), or anywhere else you write Python. Let's write our first Gradio app:

$code_hello_world

Tip: We shorten the imported name from <code>gradio</code> to <code>gr</code> for better readability of code. This is a widely adopted convention that you should follow so that anyone working with your code can easily understand it. 

Now, run your code. If you've written the Python code in a file named, for example, `app.py`, then you would run `python app.py` from the terminal.

The demo below will open in a browser on [http://localhost:7860](http://localhost:7860) if running from a file. If you are running within a notebook, the demo will appear embedded within the notebook.

$demo_hello_world

Type your name in the textbox on the left, and then press the Submit button. You should see a friendly greeting on the right.

> Tip: When developing locally, you can run your Gradio app in <strong>hot reload mode</strong>, which automatically reloads the Gradio app whenever you make changes to the file. To do this, simply type in <code>gradio</code> before the name of the file instead of <code>python</code>. In the example above, you would type: `gradio app.py` in your terminal. Learn more about hot reloading in the <a href="/guides/09_other-tutorials/developing-faster-with-reload-mode.md">Hot Reloading Guide</a>.

Reviewer: ```suggestion
Tip: When developing locally, you can run your Gradio app in <strong>hot reload mode</strong>, which automatically reloads the Gradio app whenever you make changes to the file. To do this, simply type in <code>gradio</code> before the name of the file instead of <code>python</code>. In the example above, you would type: `gradio app.py` in your terminal. Learn more about hot reloading in the <a href="/guides/developing-faster-with-reload-mode">Hot Reloading Guide</a>.
```

Author: Hmm I'm not sure if this will work when clicked from the readme. We'll need this to work on the guides though. Do you have any suggestions what to do here?

**Understanding the `Interface` Class**

You'll notice that in order to make your first demo, you created an instance of the `gr.Interface` class. The `Interface` class is designed to create demos for machine learning models which accept one or more inputs, and return one or more outputs. 

> The core `Interface` class has three core arguments:

Reviewer: ```suggestion
The `Interface` class has three core arguments:
```

- `fn`: the function to wrap a user interface (UI) around
- `inputs`: which Gradio component(s) to use for the input. The number of components should match the number of arguments in your function.
- `outputs`: which Gradio component(s) to use for the output. The number of components should match the number of return values from your function.

> The `fn` argument is very flexible -- you can pass *any* Python function that you want to wrap with a UI. In the example above, we saw a simple text-based function, but the function could be anything from music generator to a tax calculator to the prediction function of a pretrained machine learning model.

Reviewer: ```suggestion
The `fn` argument is very flexible -- you can pass *any* Python function that you want to wrap with a UI. In the example above, we saw a simple text-based function, but the function could be anything from a music generator to a tax calculator to the prediction function of a pretrained machine learning model.
```

The `input` and `output` arguments take one or more Gradio components. As we'll see, Gradio includes more than [30 built-in components](https://www.gradio.app/docs/components) (such as the `gr.Textbox()`, `gr.Image()`, `gr.Label()`, `gr.HighlightedText()`, `gr.HTML()`, etc.) that are designed for machine learning applications. For the `inputs` and `outputs` arguments, pass in the name of one of these components as a string or an instance of the class.

Tip: If your function accepts more than one argument, you can pass in a list of input components to `inputs`, with each input component corresponding to one of the arguments of the function, in order. The same holds true if your function returns more than one value: simply pass in a list of components to `outputs`. This flexibility makes the `Interface` class a very powerful way to create demos.

> We'll dive deeper into the `gr.Interface` on our series on [building Interfaces](/guides/02_building-interfaces/01_interface-state.md).

Reviewer: ```suggestion
We'll dive deeper into the `gr.Interface` on our series on [building Interfaces](/guides/interface-state).
```

Author: Same as above

## Sharing Your Demo

What good is a beautiful demo if you can't share it? Gradio lets you easily share a machine learning demo without having to worry about the hassle of hosting on a web server. Simply set `share=True` in `launch()`, and a publicly accessible URL will be created for your demo. Let's revisit our example demo,  but change the last line as follows:

```python
import gradio as gr

def greet(name):
    return "Hello " + name + "!"

demo = gr.Interface(fn=greet, inputs="textbox", outputs="textbox")
    
demo.launch(share=True)  # Share your demo with just 1 extra parameter 🚀
```

When you run this code, a public URL will be generated for your demo in a matter of seconds, something like:

👉 &nbsp; `https://a23dsf231adb.gradio.live`

Now, anyone around the world can use your Gradio application from their browser, while the machine learning model and all computation continues to run locally on your computer.

> To learn more about sharing your demo, read our dedicated guide on [sharing your Gradio application](/guides/01_getting-started/03_sharing-your-app.md).

Reviewer: ```suggestion
To learn more about sharing your demo, read our dedicated guide on [sharing your Gradio application](/guides/sharing-your-app).
```

## An Overview of Gradio

> So far, we've been discussing the `Interface` class, which is a high-level class that lets to build demos quickly with Gradio. But what else does Gradio do?

Reviewer: ```suggestion
So far, we've been discussing the `Interface` class, which is a high-level class that lets you build demos quickly with Gradio. But what else does Gradio do?
```

### Chatbots with `gr.ChatInterface`

Gradio includes another high-level class, `gr.ChatInterface`, which is specifically designed to create Chatbot UIs. Similar to `Interface`, you supply a function and Gradio creates a fully working Chatbot UI. If you're interested in creating a chatbot, you can jump straight [our dedicated guide on `gr.ChatInterface`](/guides/creating-a-chatbot-fast).

### Custom Demos with `gr.Blocks`

Gradio also offers a low-level approach for designing web apps with more flexible layouts and data flows with the `gr.Blocks` class. Blocks allows you to do things like control where components appear on the page, handle complex data flows (e.g. outputs can serve as inputs to other functions), and update properties/visibility of components based on user interaction — still all in Python. 

> You can build very custom and complex applications using `gr.Blocks()`. For example, the popular image generation [Automatic1111 Web UI](https://github.com/AUTOMATIC1111/stable-diffusion-webui) is built using Gradio Blocks. We dive deeper into the `gr.Blocks` on our series on [building with Blocks](/guides/03_building-with-blocks/01_blocks-and-event-listeners.md).

Reviewer: Why not keep a simple blocks code example? I think it helps a lot

### The Gradio Python & JavaScript Ecosystem

> So far, we've been discussing the core `gradio` Python library. But Gradio is actually so much more! Its an entire ecosystem of Python and JavaScript libraries that let you build machine learning applications, or query them programmatically, in Python or JavaScript. Here are other related parts of the Gradio ecosystem:

Reviewer: ```suggestion
That's the gist of the core `gradio` Python library, but Gradio is actually so much more! Its an entire ecosystem of Python and JavaScript libraries that let you build machine learning applications, or query them programmatically, in Python or JavaScript. Here are other related parts of the Gradio ecosystem:
```

Reviewer: You've already used "So far, we've been discussing " earlier.

Reviewer: Also sometimes you put gradio in a code block like this: `gradio` and sometimes you capitalize it: Gradio 

I guess maybe you mean when it's the core python library vs Gradio ecosystem? Even so, that hasn't been consistent

Author: Haha ok good point

> * [Gradio Python Client](/guides/08_client-libraries/01_getting-started-with-the-python-client.md) (`gradio_client`): the easiest way to query any Gradio app programmatically in Python.

Reviewer: ```suggestion
* [Gradio Python Client](/guides/getting-started-with-the-python-client) (`gradio_client`): the easiest way to query any Gradio app programmatically in Python.
```

> * [Gradio JavaScript Client](/guides/08_client-libraries/02_getting-started-with-the-js-client.md) (`@gradio/client`): the easiest way to query any Gradio app programmatically in JavaScript.

Reviewer: ```suggestion
* [Gradio JavaScript Client](/guides/getting-started-with-the-js-client) (`@gradio/client`): the easiest way to query any Gradio app programmatically in JavaScript.
```

> * [Gradio-Lite](/guides/09_other-tutorials/gradio-lite.md) (`@gradio/lite`): write Gradio apps in Python that run entirely in the browser (no server needed!), thanks to Pyodide. 

Reviewer: ```suggestion
* [Gradio-Lite](/guides/gradio-lite) (`@gradio/lite`): write Gradio apps in Python that run entirely in the browser (no server needed!), thanks to Pyodide. 
```

* [Hugging Face Spaces](https://huggingface.co/spaces): the most popular place to host Gradio applications, supports CPU instances (for free 🤗) as well as GPU instances.

## What's Next?

> Keep learning about Gradio sequentially using the Gradio Guides, which include explanations as well as example code and embedded interactive demos. Next up: 

Reviewer: ```suggestion
Keep learning about Gradio sequentially using the Gradio Guides, which include explanations as well as example code and embedded interactive demos. Next up: [Building Interfaces](/guides/interface-state)
```

# Quickstart

Gradio is an open-source Python package that allows you to quickly **build** a demo or web application for your machine learning model, API, or any arbitary Python function. You can then **share** a link to your demo or web application in just a few seconds using Gradio's built-in sharing features. *No JavaScript, CSS, or web hosting experience needed!*

<img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/gradio-guides/lcm-screenshot-3.gif" style="padding-bottom: 10px">

It just takes a few lines of Python to create a beautiful demo like the one above, so let's get started 💫

## Installation

**Prerequisite**: Gradio requires [Python 3.8 or higher](https://www.python.org/downloads/)

We recommend installing Gradio using `pip`, which is included by default in Python. Run this in your terminal or command prompt:

> ```

Author: Oh interesting, it was highlighting the word install which I thought was weird I'll play around with this
